src/fu-archive.c

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright (C) 2018 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#define G_LOG_DOMAIN				"FuArchive"
+
+#include "config.h"
+
+#include <gio/gio.h>
+#include <archive_entry.h>
+#include <archive.h>
+
+#include "fu-archive.h"
+
+/**
+ * SECTION:fu-archive
+ * @title: FuArchive
+ * @short_description: an in-memory archive decompressor
+ */
+
+struct _FuArchive {
+	GObject			 parent_instance;
+	GHashTable		*entries;
+};
+
+G_DEFINE_TYPE (FuArchive, fu_archive, G_TYPE_OBJECT)
+
+static void
+fu_archive_finalize (GObject *obj)
+{
+	FuArchive *self = FU_ARCHIVE (obj);
+
+	g_hash_table_unref (self->entries);
+	G_OBJECT_CLASS (fu_archive_parent_class)->finalize (obj);
+}
+
+static void
+fu_archive_class_init (FuArchiveClass *klass)
+{
+	GObjectClass *object_class = G_OBJECT_CLASS (klass);
+	object_class->finalize = fu_archive_finalize;
+}
+
+static void
+fu_archive_init (FuArchive *self)
+{
+	self->entries = g_hash_table_new_full (g_str_hash, g_str_equal,
+					       g_free, (GDestroyNotify) g_bytes_unref);
+}
+
+/**
+ * fu_archive_lookup_by_fn:
+ * @self: A #FuArchive
+ * @fn: A filename
+ * @error: A #GError, or %NULL
+ *
+ * Finds the blob referenced by filename
+ *
+ * Returns: (transfer none): a #GBytes, or %NULL if the filename was not found
+ **/
+GBytes *
+fu_archive_lookup_by_fn (FuArchive *self, const gchar *fn, GError **error)
+{
+	GBytes *fw;
+
+	g_return_val_if_fail (FU_IS_ARCHIVE (self), NULL);
+	g_return_val_if_fail (fn != NULL, NULL);
+	g_return_val_if_fail (error == NULL || *error == NULL, NULL);
+
+	fw = g_hash_table_lookup (self->entries, fn);
+	if (fw == NULL) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_NOT_FOUND,
+			     "no blob for %s", fn);
+	}
+	return fw;
+}
+
+static gboolean
+fu_archive_load (FuArchive *self, GBytes *blob, GError **error)
+{
+	gboolean ret = TRUE;
+	int r;
+	struct archive *arch = NULL;
+	struct archive_entry *entry;
+
+	/* decompress anything matching either glob */
+	arch = archive_read_new ();
+	archive_read_support_format_all (arch);
+	archive_read_support_filter_all (arch);
+	r = archive_read_open_memory (arch,
+				      (void *) g_bytes_get_data (blob, NULL),
+				      (size_t) g_bytes_get_size (blob));
+	if (r != 0) {
+		ret = FALSE;
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_NOT_SUPPORTED,
+			     "cannot open: %s",
+			     archive_error_string (arch));
+		goto out;
+	}
+	for (;;) {
```

Comment:
Just personal preference, I like while true better than empty for loop, it reads nicer.

Reply:
Fixed, agree.